testsuites/repository/src/main/java/org/eclipse/rdf4j/repository/RDFStarSupportTest.java

```diff
@@ -0,0 +1,131 @@
+ * are made available under the terms of the Eclipse Distribution License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *******************************************************************************/
+package org.eclipse.rdf4j.repository;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.fail;
+
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.rdf4j.model.BNode;
+import org.eclipse.rdf4j.model.IRI;
+import org.eclipse.rdf4j.model.Literal;
+import org.eclipse.rdf4j.model.Triple;
+import org.eclipse.rdf4j.model.ValueFactory;
+import org.eclipse.rdf4j.model.vocabulary.FOAF;
+import org.eclipse.rdf4j.model.vocabulary.RDF;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.Timeout;
+
+/**
+ * Test cases for RDF* support in the Repository.
+ * 
+ * @author Jeen Broekstra
+ *
+ */
+public abstract class RDFStarSupportTest {
+	/**
+	 * Timeout all individual tests after 10 minutes.
+	 */
+	@Rule
+	public Timeout to = new Timeout(10, TimeUnit.MINUTES);
+
+	private Repository testRepository;
+
+	private RepositoryConnection testCon;
+
+	private ValueFactory vf;
+
+	private BNode bob;
+
+	private BNode alice;
+
+	private BNode alexander;
+
+	private Literal nameAlice;
+
+	private Literal nameBob;
+
+	private Literal mboxAlice;
+
+	private Literal mboxBob;
+
+	private IRI context1;
+
+	private IRI context2;
+
+	@Before
+	public void setUp() throws Exception {
+		testRepository = createRepository();
+
+		testCon = testRepository.getConnection();
+		testCon.clear();
+		testCon.clearNamespaces();
+
+		vf = testRepository.getValueFactory();
+
+		// Initialize values
+		bob = vf.createBNode();
+		alice = vf.createBNode();
+		alexander = vf.createBNode();
+
+		nameAlice = vf.createLiteral("Alice");
+		nameBob = vf.createLiteral("Bob");
+
+		mboxAlice = vf.createLiteral("alice@example.org");
+		mboxBob = vf.createLiteral("bob@example.org");
+
+		context1 = vf.createIRI("urn:x-local:graph1");
+		context2 = vf.createIRI("urn:x-local:graph2");
+	}
+
+	@After
+	public void tearDown() throws Exception {
+		try {
+			testCon.close();
+		} finally {
+			testRepository.shutDown();
+		}
+	}
+
+	@Test
+	public void testAddRDFStarSubject() throws Exception {
+		Triple rdfStarTriple = vf.createTriple(bob, FOAF.NAME, nameBob);
+
+		testCon.add(rdfStarTriple, RDF.TYPE, RDF.ALT);
+
+		assertThat(testCon.hasStatement(rdfStarTriple, RDF.TYPE, RDF.ALT, false)).isTrue();
+	}
+
+	@Test
+	public void testAddRDFStarObject() throws Exception {
+		Triple rdfStarTriple = vf.createTriple(bob, FOAF.NAME, nameBob);
+
+		testCon.add(RDF.ALT, RDF.TYPE, rdfStarTriple);
+
+		assertThat(testCon.hasStatement(RDF.ALT, RDF.TYPE, rdfStarTriple, false)).isTrue();
+	}
+
+	@Test
+	public void testAddRDFStarContext() throws Exception {
+		Triple rdfStarTriple = vf.createTriple(bob, FOAF.NAME, nameBob);
```

Comment:
Just looking at this, I'm wondering how (non-)obvious the difference between `ValueFactory.createStatement` and `ValueFactory.createTriple` will be to users. Apart from making sure it's clearly explained in the javadoc, any ideas around how to make this more clear?

Reply:
If you are a complete newbie it might be unclear but if that is the case there will lots of other confusing things too. Once you realize you can't pass a Triple where a Statement is expected it should be obvious they aren't interchangeable. The multitude of methods that have "statement" in their name (hasStatement, getStatements, exportStatements, handleStatement etc.) are also a clear indication of what is expected or returned there.

If you aren't new to RDF4J it shouldn't be a problem at all.